File: 2024/problems/eeee/submissions/partially_accepted/dfs_65.py
#!/bin/python3
n = int(input())
ind = dict()
g = []
outp = []
for _ in range(n):
    dat = input().strip().split()
    if dat[0] not in ind:
        x = len(ind)
        ind[dat[0]] = x
        g.append([])
    i = ind[dat[0]]
    outp.append(i)
    for k in range(2, len(dat)):
        if dat[k][0].islower():
            continue
        if dat[k] not in ind:
            x = len(ind)
            ind[dat[k]] = x
            g.append([])
        g[i].append(ind[dat[k]])
for i in outp:
    rch = set()
    rch.add(i)
    stk = [i]
    while len(stk) > 0:
        cur = stk[-1]
        stk.pop()
        for neigh in g[cur]:
            if neigh not in rch:
                rch.add(neigh)
                stk.append(neigh)
    print(len(rch))

File: 2024/problems/eeee/submissions/partially_accepted/dfs_65.cpp
#include<bits/stdc++.h>
using namespace std;
typedef vector<int> vi;
typedef vector<vi> vvi;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);
    int n, k; string s; cin >> n;
    vvi g(n, vi());
    unordered_map<string,int> ind;
    vi outp_ind(n);
    ind.max_load_factor(0.5);
    for(int i = 0; i < n; ++i) {
        cin >> s >> k;
        if(!ind.count(s)) {
            int x = ind.size();
            ind[s] = x;
        }
        int y = ind[s];
        outp_ind[i] = y;
        for(int j = 0; j < k; ++j) {
            cin >> s;
            if(islower(s[0])) continue;
            if(!ind.count(s)) {
                int x = ind.size();
                ind[s] = x;
            }
            g[y].push_back(ind[s]);
        }
    }
    for(int i : outp_ind) {
        vector<bool> rch(n, false); rch[i] = true;
        vector<int> stk; stk.push_back(i);
        int ans = 1;
        while(stk.size() > 0) {
            int cur = stk.back(); 
            stk.pop_back();
            for(int neigh : g[cur]) {
                if(!rch[neigh]) {
                    rch[neigh] = true;
                    stk.push_back(neigh);
                    ans++;
                }
            }
        }
        cout << ans << '\n';
    }
}
